src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -1579,79 +1453,74 @@ private CorInfoCallConvExtension getUnmanagedCallConv(CORINFO_METHOD_STRUCT_* me
+            }
+
+            return ToCorInfoCallConvExtension(callingConventions, out suppressGCTransition);
         }
 
         private CorInfoCallConvExtension GetUnmanagedCallConv(MethodSignature signature, out bool suppressGCTransition)
         {
-            suppressGCTransition = false;
-            switch (signature.Flags & MethodSignatureFlags.UnmanagedCallingConventionMask)
+            return ToCorInfoCallConvExtension(signature.GetStandaloneMethodSignatureCallingConventions(), out suppressGCTransition);
+        }
+
+        private CorInfoCallConvExtension ToCorInfoCallConvExtension(CallingConventions callConvs, out bool suppressGCTransition)
+        {
+            CorInfoCallConvExtension result;
+            switch (callConvs & CallingConventions.CallingConventionMask)
             {
-                case MethodSignatureFlags.None:
-                    ThrowHelper.ThrowInvalidProgramException();
-                    return CorInfoCallConvExtension.Managed;
-                case MethodSignatureFlags.UnmanagedCallingConventionCdecl:
-                    return CorInfoCallConvExtension.C;
-                case MethodSignatureFlags.UnmanagedCallingConventionStdCall:
-                    return CorInfoCallConvExtension.Stdcall;
-                case MethodSignatureFlags.UnmanagedCallingConventionThisCall:
-                    return CorInfoCallConvExtension.Thiscall;
-                case MethodSignatureFlags.UnmanagedCallingConvention:
-                    if (TryGetUnmanagedCallingConventionFromModOpt(signature, out CorInfoCallConvExtension callConvMaybe, out suppressGCTransition))
-                    {
-                        return callConvMaybe;
-                    }
-                    else
-                    {
-                        return (CorInfoCallConvExtension)PlatformDefaultUnmanagedCallingConvention();
-                    }
+                case CallingConventions.Cdecl:
+                    result = CorInfoCallConvExtension.C;
+                    break;
+                case CallingConventions.Stdcall:
+                    result = CorInfoCallConvExtension.Stdcall;
+                    break;
+                case CallingConventions.Thiscall:
+                    result = CorInfoCallConvExtension.Thiscall;
+                    break;
+                case CallingConventions.Fastcall:
+                    result = CorInfoCallConvExtension.Fastcall;
+                    break;
+                case CallingConventions.Unmanaged:
+                    result = _compilation.TypeSystemContext.Target.IsWindows ?
+                        CorInfoCallConvExtension.Stdcall : CorInfoCallConvExtension.C;
+                    break;
                 default:
                     ThrowHelper.ThrowInvalidProgramException();
-                    return CorInfoCallConvExtension.Managed;
+                    result = CorInfoCallConvExtension.Managed;
```

Comment:
We set even though we are throwing?

Reply:
It is to make the compiler happy. The compiler does not see that we are throwing.

Reply:
I think it would sense then to make that clear:

```csharp
throw new Exception("Defer to throwing helper");
```

Reply:
Setting things made me assume the helper was conditional for some reason.

Reply:
The dummy assignment in situations like this is the common practice in the repo. You can find number of places that do this (one example https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs#L153).

I agree that it is not obvious what's going on when you see it for the first time.

Reply:
I've added a comment that this is unreached.